cpu/stm32/periph/flashpage.c

```diff
@@ -253,3 +319,54 @@ void flashpage_write(void *target_addr, const void *data, size_t len)
     }
 #endif
 }
+
+#if !(defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || \
+      defined(CPU_FAM_STM32F7))
+void flashpage_write_page(unsigned page, const void *data)
+{
+    flashpage_erase(page);
+
+    /* WRITE sequence */
+    if (data != NULL) {
+        flashpage_write(flashpage_addr(page), data, FLASHPAGE_SIZE);
+    }
+}
+#endif
+
+#if defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || \
+    defined(CPU_FAM_STM32F7)
+size_t flashpage_size(unsigned page)
+{
+    if (page < 4) {
+        return 16 * 1024;
+    }
+    else if (page == 4) {
+        return 64 * 1024;
+    }
+    else {
+        return 128 * 1024;
+    }
+}
+
+void *flashpage_addr(unsigned page)
+{
+    if (page == 0) {
+        return (void*)CPU_FLASH_BASE;
+    }
+    return (void*)((intptr_t)flashpage_addr(page - 1) + flashpage_size(page - 1));
```

Comment:
Will the compiler be able to resolve this recursion?
I'm worried about the stack…

Reply:
absolutely not, and I have a more elegant line to replace this with.

Reply:
We could have a generic 

```C
void *flashpage_addr(unsigned page)
{
    uintptr_t addr = 0;

    while (page) {
        addr += flashpage_size(--page);
    }

    return addr;
}
```

Reply:
Rewrote this to use your suggestion